nanostack-event-loop/eventOS_event.h

```diff
@@ -19,6 +19,173 @@
+ * Rest of the fields are optional, and used only by the receiving callback. So you have different options to
+ * deliver data to a receiving tasklet. The structure is copied by the event system, so temporary storage may be used,
+ * and the structure may be freed after it has been pushed into event system.
+ *
+ * \code
+ *      // Send the event
+ *      arm_event_t e = {
+ *       .receiver = my_eventhandler_id,
+ *       .event_type = MY_EVENT
+ *      };
+ *
+ *      if (eventOS_event_send(e) != 0) {
+ *       // fail
+ *      }
+ * \endcode
+ *
+ * Where required, event system allows you to delay the event propagation.
+ *
+ * \code
+ *      // Wait 3 seconds before the event
+ *      #define MY_DELAY_MS 3000
+ *
+ *      arm_event_t e = {
+ *          .receiver = my_eventhandler_id,
+ *          .event_type = MY_EVENT
+ *      };
+ *
+ *      uint32_t delay = eventOS_event_timer_ms_to_ticks(MY_DELAY_MS);
+ *      eventOS_event_send_after(e, delay);
+ * \endcode
+ *
+ * \sa eventOS_event.h
+ * \sa eventOS_event_send_at
+ * \sa eventOS_event_send_in
+ * \sa eventOS_event_send_after
+ * \sa eventOS_event_send_every
+ *
+ * <h3>Pre-allocated events</h3>
+ *
+ * Two options are provided to limit the heap usage. First option is to use reccuring events with eventOS_event_send_every(),
+ * so your event is only allocated once. This allows you to create application that does not use heap after initialization phase.
+ *
+ * Second option is to use pre-allocated or statically allocated event structure. In this model you create a space for
+ * \ref arm_event_storage structure and send events using eventOS_event_send_user_allocated() call.
+ * \code
+ *   static bool pending = false;
+ *   static arm_event_storage_t e;
+ *   static int8_t foo_tasklet_id;
+ *
+ *   void notify_foo()
+ *   {
+ *       if (!pending) {
+ *           pending = true;
+ *           e.data.receiver = foo_tasklet_id;
+ *           e.data.type = MY_EVENT;
+ *           eventOS_event_send_user_allocated(&e);
+ *      }
+ *   }
+ *
+ *   void foo_event_handler(arm_event_t *e)
+ *   {
+ *       pending = false;
+ *       // ...
+ *   }
+ *
+ * \endcode
+ *
+ * <h3>Initialization</h3>
+ *
+ * Event system does not use malloc(), free() or any system heap directly, but uses nsdynmemLIB.h library instead.
+ * Event system must first be initialized by callind eventOS_scheduler_init(). This is usually done just after ns_dyn_mem_init() call.
+ * Where porting is already provided, these both are initialized in function called ns_hal_init().
+ *
+ * After initialization, you can start the event loop by calling eventOS_scheduler_run() which will never return. This is usually
+ * end of the `main()` function.
+ *
+ * \code
+ * extern void my_event_handler(arm_event_t *e);
+ * extern int8_t my_eventhandler_id;
+ *
+ * void main(void)
+ * {
+ *     ns_dyn_mem_init(NULL, HEAP_SIZE, NULL, NULL);
+ *     my_eventhandler_id = eventOS_event_handler_create(my_event_handler, INITIALIZATION_EVENT);
```

Comment:
Don't you need to call ns_hal_init() or eventOS_scheduler_init() too?

Reply:
Goot catch.. I did talk about `eventOS_scheduler_init()` but forget to add it here.

`ns_hal_init()` is not part of our API, but instead a name usually used by our platform porting, which tend to just be dynmem init and eventloop init.